metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -2305,11 +2307,34 @@ public CompletableFuture<UpdateFeaturesResponseData> updateFeatures(
         }).thenApply(result -> {
             UpdateFeaturesResponseData responseData = new UpdateFeaturesResponseData();
             responseData.setResults(new UpdateFeaturesResponseData.UpdatableFeatureResultCollection(result.size()));
-            result.forEach((featureName, error) -> responseData.results().add(
-                new UpdateFeaturesResponseData.UpdatableFeatureResult()
-                    .setFeature(featureName)
-                    .setErrorCode(error.error().code())
-                    .setErrorMessage(error.message())));
+            Optional<Entry<String, ApiError>> errorEntry = Optional.empty();
+            if (context.requestHeader().requestApiVersion() > 1) {
+                Stream<Entry<String, ApiError>> errorEntries = result.entrySet().stream().filter(entry ->
+                        !entry.getValue().error().equals(Errors.NONE));
+                errorEntry = errorEntries.findFirst();
+            }
+                
+            if (errorEntry.isPresent()) {
+                String errorFeatureName = errorEntry.get().getKey();
+                ApiError topError = errorEntry.get().getValue();
+                String errorString = errorFeatureName + ":" + topError.error().exceptionName() + " (" + topError.message() + ")";
+                responseData.setErrorCode(Errors.INVALID_UPDATE_VERSION.code());
```

Comment:
The KIP specified this error code. We can change this if we think this should not be the case. 
> The UpdateFeaturesReponse will contain a top level INVALID_UPDATE_VERSION  error if any feature fails to update and no updates will persist if a single feature in the requrest fails validation. The same INVALID_UPDATE_VERSION  error will be specified in UpdatableFeatureResult.ErrorCode for the features that failed. This will be accompanied with UpdatableFeatureResult.ErrorMessage an error message explaining the dependency.

Reply:
Ok, but is ControllerResult.response returned from `featureControl.updateFeatures` supposed to contain the error code for the client?

Reply:
I included it there so the error message could contain the "real" error.

Reply:
@junrao I guess to summarize the conversation -- does it make sense to simply return whatever error occurred as the top level error and not invalid update version? I would think we would still want the specific feature in the response message so it can be communicated to the user.

Reply:
It's fine to convert arbitrary server errors to just `INVALID_UPDATE_VERSION`. I was just wondering where the conversion happens. It seems that we already did that conversion in `FeatureControlManager.updateFeature()`. Do we need to do the same conversion here again?

Reply:
Ah. I see what you are saying. So we should just take the error and not further convert it. The main thing that is important is to also include the feature name.